### ISSUE TITLE
[Question] - Missing warning?

### ISSUE THREAD

**Issue by LionyxML** [thefrontside/frontmacs#156]
Hello there!

Thanks for bringing me much closer to never open VSCode again. I used to use TS-Tide along with other modes but I could not rely on my custom configs (mostly my fault, obviously). Thank you a lot for the heavy lifting work!

I kind of have a doubt but I am not sure it is related to `frontmacs` it self, if anyone is willing to help...

I noticed that some warnings from VSCode are not displayed when using `frontside-javascript` mode.

Example:

Consider this piece of code:

```
  const keyDown = (
    e: React.KeyboardEventHandler<HTMLInputElement> | undefined
  ): void => {
    if (e.code === "Backspace" && !value) {
      e.preventDefault();
    }
```
On VSCode I am warned this object might be undefined so a bunch of bad things might happen:
<img width="739" alt="image" src="https://user-images.githubusercontent.com/16169950/156891415-3632eaee-03ea-41c3-ad25-6cc30b0166a7.png">

On emacs with `frontside-javascript` mode on the same file, no warning is displayed:
<img width="591" alt="image" src="https://user-images.githubusercontent.com/16169950/156891439-52adc196-d418-4d75-b7e8-8bcf94d14d43.png">

What should I fix on emacs to have these kinds of warnings displaying?

**Comment by cowboyd** [comment]
> Thanks for bringing me much closer to never open VSCode again. 

We aim to please 😄 

Hm.... It should use `flycheck` to display these errors (it does, in fact annotate my buffers), so clearly there is a bug somewhere, either in `frontside-javascript`, What do you see if you run

```lisp
(flycheck-list-errors)
```

That should show all the errors you expect to see annotating the code in your buffer inside a table. To verify your `flycheck` is working properly, you can run

```eslisp
(flycheck-verify-setup)
```

It should look something like this:

```
Syntax checkers for buffer example.ts in typescript-mode:

First checker to run:

  typescript-tide
    - may enable:        yes
    - may run:           t
    - Typescript server: running
    - Tide mode:         enabled
    - next checkers:     javascript-eslint

Checkers that may run as part of the first checker's chain:

  javascript-eslint
    - may enable:  yes
    - executable:  Found at /Users/cowboyd/Code/@frontmacs/example/node_modules/.bin/eslint
    - config file: found

Checkers that could run if selected:

  typescript-tslint  select
    - may enable:         yes
    - executable:         Found at /Users/cowboyd/Code/@frontmacs/example/node_modules/.bin/tslint
    - configuration file: Not found

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------
```

Let me know what you find!

**Comment by LionyxML** [comment]
Thanks again!

`flycheck-list-errors` shows no errors in the file.

Now with `flycheck-verify-setup` this is what I get:

```
Syntax checkers for buffer ****.tsx in web-mode:

First checker to run:

  javascript-eslint
    - may enable:  yes
    - executable:  Found at ..../node_modules/.bin/eslint
    - config file: found

Checkers that are compatible with this mode, but will not run until properly configured:

  tsx-tide (manually disabled) enable
    - may enable:        yes
    - may run:           t
    - Typescript server: running
    - Tide mode:         enabled

  ember-template (automatically disabled) reset
    - may enable:         no
    - executable:         Not found
    - configuration file: Not found

  handlebars (automatically disabled) reset
    - may enable: no
    - may run:    nil
    - executable: Not found

  jsx-tide (manually disabled) enable
    - may enable:        yes
    - may run:           nil
    - Typescript server: running
    - Tide mode:         enabled

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 32snapshot (package: 20210825.1804)
Emacs version:    27.2
System:           arm-apple-darwin21.1.0
Window system:    ns
```

So I enabled `tsx-tide` and tada:
![image](https://user-images.githubusercontent.com/16169950/157136503-e995ff7b-da6f-47d4-a794-e63013a6123e.png)

But now, if enable `tsx-tide` I cant see "unexpected any", only when I enable `javascript-eslint`:
![image](https://user-images.githubusercontent.com/16169950/157137100-3b81f8c3-dd48-4648-b050-c5d0d0e829ef.png)

It seems I can get `typescript` OR `eslint`, not both.

Also, did you see it says I am in "web-mode" is it expected?

**Comment by cowboyd** [comment]
@LionyxML Ok, so you definitely should see it in your linting errors. Is there any way you could reproduce this in a public repo that I could check out? There is so much at play including package.json, tsconfig.json, etc....

**Comment by LionyxML** [comment]
Thanks again! :D and I am sorry flooding you, but I tried to show everything as detailed as I could.

About the code, I sure can update something! Here's a repo that reproduces this behavior: [https://github.com/LionyxML/frontmacs_test_case/](https://github.com/LionyxML/frontmacs_test_case/)

Everything but the `src/` is the same as in the real project (package.json, yarn.lock, etc).

If I open `src/App.tsx` on vscode I get:
<img width="736" alt="image" src="https://user-images.githubusercontent.com/16169950/157352580-4062c096-28b0-47b8-ba16-820c19383b3f.png">
We can see warnings on lines: 6, 13 and 19.

Running the app with `yarn start` shows this same warnings on dev server (localhost:3000):
<img width="989" alt="image" src="https://user-images.githubusercontent.com/16169950/157352732-dfff2202-7163-4c0f-b10e-06df369dcd90.png">

On emacs, just opening the same file gives me:
<img width="683" alt="image" src="https://user-images.githubusercontent.com/16169950/157352844-c962cc3f-1109-4011-b902-6987f090ff3d.png">
And you see the  "missing warnings" missing.

In this situation if I run `flycheck-verify-setup`this is shown:
<img width="822" alt="image" src="https://user-images.githubusercontent.com/16169950/157352948-c17d4738-4e39-4448-82e9-22d50845cbdc.png">

If I click on select near `tsx-tide`:

<img width="824" alt="image" src="https://user-images.githubusercontent.com/16169950/157353035-46acd2b3-7fe4-47d4-9fdf-d792898617ae.png">

And then emacs ONLY shows these warnings:
<img width="692" alt="image" src="https://user-images.githubusercontent.com/16169950/157353077-f469034c-a57b-490b-8fc3-493a1bbc8c95.png">

I can't manage to have all warnings like "stock" vscode. It is eather js-tide OR eslint, never `both`.

Thanks again for your help!




<blockquote><img src="https://opengraph.githubassets.com/bfb07f3a9f7ffdc9eccf3197145824f0fb03f7e6bfcae9303cd8f58b0c7bef7e/LionyxML/frontmacs_test_case" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/LionyxML/frontmacs_test_case">GitHub - LionyxML/frontmacs_test_case: A Typescript / ReactJS repo to explore some of frontmacs features</a></strong></div><div>A Typescript / ReactJS repo to explore some of frontmacs features - GitHub - LionyxML/frontmacs_test_case: A Typescript / ReactJS repo to explore some of frontmacs features</div></blockquote>

**Comment by cowboyd** [comment]
> Thanks again! :D and I am sorry flooding you, but I tried to show everything as detailed as I could.

No worries, and thank you for following through! The good news is that your repo reproduces it for me as well, so looks like you've definitely found a bug in `frontside-javascript`.

I'm having a look now.

**Comment by cowboyd** [comment]
Web mode  _is_ expected for `.tsx` files since there is no typescript equivalent of `rjsx-mode` and `typescript-mode` does not support JSX. It's not ideal, but it's the most common answer on web, and if anything, `frontside-javascript` is about not having to spend 6 hours googling just to get all the javascript edge cases to work.

Ok, here's the problem. The `tsx-tide`checker was explicitly disabled because of the problem you describe: you either get the `javascript-eslint` errors, or you get the `tsx-tide` errors, but not both, and based on such a compromise, I had chosen the eslint errors because those were most "blocking" in development flow. In other words, your test suite is going to catch your type errors, but you may not be aware of linting errors until you run your code in CI. 

(In normal `.ts` files, this is not a problem because `typescript-mode` highlights syntax errors)

https://github.com/thefrontside/frontmacs/blob/release/packages/javascript/frontside-javascript.el#L177-L179

However, it appears that flycheck _does_ actually support running multiple checkers via checker chains https://www.flycheck.org/en/latest/user/syntax-checkers.html#function-flycheck-add-next-checker

I've tested this out by adding enabling all of the checkers, and then chaining together the two checkers with a call to 

```elisp
(flycheck-add-next-checker 'tsx-tide 'javascript-eslint)
```

This results in the correct error highlighting, and you can see that there are errors in flycheck for _both_ `tex-tide` and `javascript-eslint`

![image](https://user-images.githubusercontent.com/4205/157411279-c2bf1495-9806-4ada-8a46-c16c785a1976.png)


Until I can get a PR put together, this is how I have my javascript configured in my init file:

```elisp
(use-package frontside-javascript
  :init (frontside-javascript)
  :config
  (advice-add
   'frontside-javascript--tsx-web-mode-hook :after
   (lambda ()
     (setq flycheck-disabled-checkers nil)
     (flycheck-add-next-checker 'tsx-tide 'javascript-eslint))))
```

It may take me a bit to get to this, but if you want to pair on it at some time, I'd be happy to fix it with you to get a release faster.

<blockquote><img src="https://opengraph.githubassets.com/d9fed3747e9c655a937355ec89caaaceea677bced0f0f34e6f48bc4fafed3ee6/thefrontside/frontmacs" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/thefrontside/frontmacs">frontmacs/frontside-javascript.el at release · thefrontside/frontmacs</a></strong></div><div>Package-based, web-centric, customizable, awesome-by-default, acceptance-tested Emacs distribution - frontmacs/frontside-javascript.el at release · thefrontside/frontmacs</div></blockquote>
<blockquote><img src="../_static/favicon.ico" width="48" align="right"><div><strong><a href="https://www.flycheck.org/en/latest/user/syntax-checkers.html#function-flycheck-add-next-checker">Syntax checkers — Flycheck 32-cvs documentation</a></strong></div></blockquote>

**Comment by LionyxML** [comment]
Thanks again! This configuration indeed works!

I think I might be way too weak in  my elisp skills to help. But I tried to do something with a PR (above), it compiles and runs with no errors in my system and allows me a workflow really close of what I get on *cof* vscode.

**Comment by cowboyd** [comment]
> I think I might be way too weak in my elisp skills to help.

That's where we all started! 

> But I tried to do something with a PR (above), it compiles and runs with no errors in my system and allows me a workflow really close of what I get on cof vscode.

Awesome! I'll try and write a test case for it so that we can get it merged. What other things are you missing from VSCode?

**Comment by cowboyd** [comment]
closed by #157
